RecoParticleFlow/PFProducer/plugins/SimPFProducer.cc

```diff
@@ -0,0 +1,295 @@
+  
+  virtual void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+  
+private:
+  // parameters
+  const double superClusterThreshold_;
+  
+  // inputs
+  const edm::EDGetTokenT<edm::View<reco::Track> > tracks_;
+  const edm::EDGetTokenT<edm::View<reco::Track> > gsfTracks_;
+  const edm::EDGetTokenT<TrackingParticleCollection> trackingParticles_;
+  const edm::EDGetTokenT<SimClusterCollection> simClustersTruth_;
+  const edm::EDGetTokenT<CaloParticleCollection> caloParticles_;
+  const edm::EDGetTokenT<std::vector<reco::PFCluster> > simClusters_;
+  // tracking particle associators by order of preference
+  const std::vector<edm::EDGetTokenT<reco::TrackToTrackingParticleAssociator> > associators_;     
+};
+
+DEFINE_FWK_MODULE(SimPFProducer);
+
+namespace {
+  
+  template<typename T>
+  uint64_t hashSimInfo(const T& simTruth,size_t i = 0) {
+    uint64_t evtid = simTruth.eventId().rawId();
+    uint64_t trackid = simTruth.g4Tracks()[i].trackId();
+    return ( (evtid << 3) + 23401923 ) ^ trackid ;
+  };
+}
+
+SimPFProducer::SimPFProducer(const edm::ParameterSet& conf) :
+  superClusterThreshold_( conf.getParameter<double>("superClusterThreshold") ),
+  tracks_(consumes<edm::View<reco::Track> >( conf.getParameter<edm::InputTag>("trackSrc") ) ),
+  gsfTracks_(consumes<edm::View<reco::Track> >( conf.getParameter<edm::InputTag>("gsfTrackSrc") ) ),
+  trackingParticles_(consumes<TrackingParticleCollection>( conf.getParameter<edm::InputTag>("trackingParticleSrc") ) ),
+  simClustersTruth_(consumes<SimClusterCollection>( conf.getParameter<edm::InputTag>("simClusterTruthSrc") ) ),
+  caloParticles_(consumes<CaloParticleCollection>( conf.getParameter<edm::InputTag>("caloParticlesSrc") ) ),
+  simClusters_(consumes<std::vector<reco::PFCluster> >( conf.getParameter<edm::InputTag>("simClustersSrc") ) ),
+  associators_( edm::vector_transform( conf.getParameter<std::vector<edm::InputTag> >("associators"), [this](const edm::InputTag& tag){ return this->consumes<reco::TrackToTrackingParticleAssociator>(tag); } ) )
+{
+  produces<reco::PFBlockCollection>();
+  produces<reco::SuperClusterCollection>("perfect");
+  produces<reco::PFCandidateCollection>();
+}
+
+void SimPFProducer::produce(edm::StreamID, edm::Event& evt, const edm::EventSetup& es) const {  
+  //get associators
+  std::vector<edm::Handle<reco::TrackToTrackingParticleAssociator> > associators;
+  for( const auto& token : associators_ ) {
+    associators.emplace_back();
+    auto& back = associators.back();
+    evt.getByToken(token,back);
+  }
+  
+  //get track collections
+  edm::Handle<edm::View<reco::Track> > TrackCollectionH;
+  evt.getByToken(tracks_, TrackCollectionH);
+  const edm::View<reco::Track>& TrackCollection = *TrackCollectionH;
+
+  /*
+  edm::Handle<edm::View<reco::Track> > GsfTrackCollectionH;
+  evt.getByToken(gsfTracks_, GsfTrackCollectionH);
+  const edm::View<reco::Track>& GsfTrackCollection = *GsfTrackCollectionH;
+  */
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant